### PR TITLE
closes #7 database create user model in db

### LINKED ISSUE #7: [Database]: Create User Model in DB
### Description

Database Table - Users

| :x: New route |  :x: An existing route | :white_check_mark: Design/Build/Database |
| --- | --- | --- |

### Feature description

Create the User Model in Repo using PostgreSQL DB

1. Needs the following attributes:
    - [ ] Email - String
    - [ ] Password Digest - String
    - [ ] Timestamps (created/updated) - DateTime
2. Setup Password encryption
3. Test for validations
4. Test happy / sad / and edge cases
5. Add Documentation as and if needed

**Remember:** Use GitHub branch creation for naming
Image

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by wanderlust-create**
## What Type of PR is this? 
- [x] 🎁 Feature
- [x]  ✅ Test
- [x] 📊 Changes DB
## Describe your changes 
  <sup>Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.</sup>
created User model with validation testing 
## Issue number and link 
<sup>Only PR's for existing issue tickets or features for this sprint will be approved. Issue _MUST_ be linked for review.</sup>
Closes #7

## Questions for reviewers?
<sup>Please note anything you would like specific feedback on in your code.</sup>

## Checklist before requesting a review 

| Repo Contribution Standards| Complete |
| ----- | ---------------- |
| I have performed a self-review of my code | <ul><li> - [x] **Yes** </li><li> [ ] **No**</li> |
| My code follows the style guidelines of this project| <ul><li> - [x] **Yes** </li><li> [ ] **No**</li> |
| I have made corresponding changes to the documentation (Note: Documentation _must_ be up to date for merge)| <ul><li> [ ] **Yes** </li><li> - [x] **No**: there is no documentation yet to update</li> |

| Change Type | Complete |
| ----- | ---------------- |
| Non-breaking change which fixes an issue | <ul><li> - [x] **Yes** </li><li> [ ] **No**</li> |
| Breaking change (fix or feature that would cause existing functionality to not work as expected) | <ul><li> [ ] **Yes** </li><li> - [x] **No**</li> | 

| Testing | Complete |
| ----- | ---------------- |
| I have added tests that prove my fix is effective or that my feature works | <ul><li> - [x] **Yes** </li><li> [ ] **No**</li> |
| New and existing tests pass locally with my changes | <ul><li> - [x] **Yes** </li><li> [ ] **No**</li> |

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  validates :email, presence: true, format: { with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i, on: :create }
+  validates :email, uniqueness: { case_sensitive: false }
+  validates :password_digest, presence: true
```
Comment (aleish-m) [inline app/models/user.rb:4]: Not sure if we can do this validation and the `has_secure_password` or if this is the right way to have this type of validation... 
The `has_secure_password` validator does **ALOT**, so it might be doing this already. I would double-check both on [Rails docs](https://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html) and the bcrypt docs/lesson. I remember something specific about how this validation has to work to stay secure and not cause in-app issues. 

I'll look into it more too. Obviously, I need to refresh my memory also!

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+require 'rails_helper'
+
+RSpec.describe User do
+  describe 'validations' do
+    it { is_expected.to validate_presence_of(:email) }
+    it { is_expected.to validate_uniqueness_of(:email).case_insensitive }
+    it { is_expected.to validate_presence_of(:password_digest) }
+  end
+end
```
[the remark this answers is not in the document]
Comment (aleish-m) [reply spec/models/user_spec.rb:9]: example of password tests from mod 3 viewing party project when bcrypt was introduced:

```Ruby
it { should have_secure_password}

    it "stores secured paswords as password_digest not as password" do
      user = User.create(name: 'Meg', email: 'meg@test.com', password: 'password123', password_confirmation: 'password123')
      expect(user).to_not have_attribute(:password)
      expect(user.password_digest).to_not eq('password123')
    end
```

I would probably still leave your `presence_of(:password_digest)` or add a line to the above test for `expect(user).to_have_attribute(:password_digest)` to show we take in a password but save it as password digest and encrypt it.

**Review by aleish-m — CHANGES_REQUESTED** [review]
A couple of things were noted. Let me know if you have questions or find cool answers, reference docs, etc.

PS. Thanks for getting this up so quickly! I was able to fix the things in workflows (CI and auto-assign) that I needed this PR to accomplish 🥳

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  validates :email, presence: true, format: { with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i, on: :create }
```
[the remark this answers is not in the document]
Comment (wanderlust-create) [reply app/models/user.rb:2]: 3️⃣ since we are only collecting an email and password I think if they use a different email, it would register as a whole new account. I know that is the case for me. I have to make sure I sign in with the same email to get access to the data I created earlier (I have both a personal and teaching email that I use). 

4️⃣ I  choose to use Regex to validate the email because it should not need any maintenance. I read this thread and choose it as the best option: https://stackoverflow.com/questions/38611405/email-validation-in-ruby-on-rails
It is also listed as one of the Ruby/ Rails validators on this link: https://api.rubyonrails.org/v5.1/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates
Do you want me to add these links as comments to the code?

Side note: we can definitely pass this through to the FE when we build the Post path. Right now, there is no create user API route as this is only the database creation. Agreed, when we create the route we should return the username.

File: db/migrate/20230721213350_add_index_to_users_email.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToUsersEmail < ActiveRecord::Migration[7.0]
+  def change
+    add_index :users, :email, name: "index_users_on_email", unique: true
```
[the remark this answers is not in the document]
Comment (wanderlust-create) [reply db/migrate/20230721213350_add_index_to_users_email.rb:3]: I needed to do this to check that the email is unique.

File: spec/factories/users.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :user do
+    email { 'MyString' }
```
[the remark this answers is not in the document]
Comment (wanderlust-create) [reply spec/factories/users.rb:3]: I did not do this. I think it is working as it should! Yea!

File: spec/factories/users.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+    password_digest { 'MyString' }
```
[the remark this answers is not in the document]
Comment (wanderlust-create) [reply spec/factories/users.rb:4]: Yes, I did not write any of this code. It was auto-created when I used rails g model User cli.

File: spec/models/user_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (wanderlust-create) [reply spec/models/user_spec.rb:9]: let's have a chat when you get back about how we expect the passwords to look. I only added happy path validations to get it up and running asap. We can definitely create more specific requirements.

File: app/models/user.rb (hunk 1 above)
Comment (wanderlust-create) [reply app/models/user.rb:4]: When I wrote my rails app in mod 3 this is how I created it. I basically used that project to show me how to set it up. Let's have a discussion about the specifics you are wanting to know. I was under the impression that we were mostly interested in getting OAuth-type verification set up for this long run and this was just to get the app running for now. If you have specific concerns, we can definitely look into it more.

File: app/models/user.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (aleish-m) [reply app/models/user.rb:2]: I think you make a good point that if we are going to do anything with updating email addresses, that is a discussion and separate issue down the road. Let's put some blinders on and not feature creep on this issue right now.

I think those are great resources to have used as a reference so I'm glad you added them to the comments for the PR for later reference. What I'm talking about for code comments is a comment detailing exactly what the format that the regex is looking for in case we need to change what a valid format looks like down the road or we have issues with the format validations. making it easier to debug, break down, and understand exactly what the regex code says since it is unfamiliar and very specific. (eg: `This checks that there is at least one character of any type non-case sensitive followed by an @ symbol ...`)

File: db/migrate/20230721213350_add_index_to_users_email.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (aleish-m) [reply db/migrate/20230721213350_add_index_to_users_email.rb:3]: https://guides.rubyonrails.org/active_record_validations.html#uniqueness:~:text=2.12%20uniqueness,in%20that%20attribute.

File: spec/factories/users.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (aleish-m) [reply spec/factories/users.rb:3]: Added Faker generator to User Factory for both email and password, along with a comment on available password arguments for future reference.

File: spec/factories/users.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (aleish-m) [reply spec/factories/users.rb:4]: Added Faker generator to User Factory for both email and password, along with a comment on available password arguments for future reference.

File: spec/models/user_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (aleish-m) [reply spec/models/user_spec.rb:9]: - [Documentation](https://api.rubyonrails.org/v5.1/classes/ActiveModel/SecurePassword/ClassMethods.html) for what `has_secure_password` validates
- Added Happy and Sad path testing for email formatting validations
   - Added Rubocop configuration:
      - Allow up to 2 Expects in 1 test (and disable for one test that uses 4)
      - Disable describe blocks Nesting depth for this file
      - Allow tests to be 8 lines long

File: app/models/user.rb (hunk 1 above)
Comment (aleish-m) [reply app/models/user.rb:10]: corrected model validation

**Review by aleish-m — DISMISSED** [review]

**Review by aleish-m — APPROVED** [review]
reviewed changes and completed errors with merging the main and this branch for work flow and issue template files.
